core/clustersmngr/factory.go

```diff
@@ -0,0 +1,306 @@
+	return &clientsFactory{
+		hubClient:          hubClient,
+		clustersFetcher:    fetcher,
+		nsChecker:          nsChecker,
+		clusters:           &Clusters{},
+		clustersNamespaces: &ClustersNamespaces{},
+		usersNamespaces:    &UsersNamespaces{Cache: ttlcache.New(24 * time.Hour)},
+		log:                logger,
+	}
+}
+
+func (cf *clientsFactory) Start(ctx context.Context) {
+	go cf.watchClusters(ctx)
+	go cf.watchNamespaces(ctx)
+}
+
+func (cf *clientsFactory) watchClusters(ctx context.Context) {
+	if err := wait.PollImmediateInfinite(watchClustersFrequency, func() (bool, error) {
+		if err := cf.UpdateClusters(ctx); err != nil {
+			return false, err
+		}
+
+		return true, nil
+	}); err != nil {
+		cf.log.Error(err, "failed pooling clusters")
+	}
+}
+
+func (cf *clientsFactory) UpdateClusters(ctx context.Context) error {
+	clusters, err := cf.clustersFetcher.Fetch(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch clusters: %w", err)
+	}
+
+	cf.clusters.Set(clusters)
+
+	return nil
+}
+
+func (cf *clientsFactory) watchNamespaces(ctx context.Context) {
+	if err := wait.PollImmediateInfinite(watchNamespaceFrequency, func() (bool, error) {
+		if err := cf.UpdateNamespaces(ctx); err != nil {
+			return false, err
+		}
+
+		return false, nil
+	}); err != nil {
+		cf.log.Error(err, "failed pooling namespaces")
+	}
+}
+
+func (cf *clientsFactory) UpdateNamespaces(ctx context.Context) error {
+	clients, err := clientsForClusters(cf.clusters.Get())
+	if err != nil {
+		cf.log.Error(err, "failed to create clients for", "clusters", cf.clusters.Get())
+		return err
+	}
+
+	wg := sync.WaitGroup{}
+
+	for clusterName, c := range clients {
+		wg.Add(1)
+
+		go func(clusterName string, c client.Client) {
+			defer wg.Done()
+
+			nsList := &v1.NamespaceList{}
+
+			if err := c.List(ctx, nsList); err != nil {
+				cf.log.Error(err, "failed listing namespaces", "cluster", clusterName)
+			}
+
+			cf.clustersNamespaces.Set(clusterName, nsList.Items)
+		}(clusterName, c)
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+func (cf *clientsFactory) GetImpersonatedClient(ctx context.Context, user *auth.UserPrincipal) (Client, error) {
+	pool := NewClustersClientsPool()
+
+	for _, cluster := range cf.clusters.Get() {
+		if err := pool.Add(ClientConfigWithUser(user), cluster); err != nil {
+			return nil, fmt.Errorf("failed adding cluster client to pool: %w", err)
+		}
+	}
+
+	return NewClient(pool, cf.userNsList(ctx, user)), nil
+}
+
+func (cf *clientsFactory) GetServerClient(ctx context.Context) (Client, error) {
+	pool := NewClustersClientsPool()
+
+	for _, cluster := range cf.clusters.Get() {
+		if err := pool.Add(restConfigFromCluster, cluster); err != nil {
+			return nil, fmt.Errorf("failed adding cluster client to pool: %w", err)
+		}
+	}
+
+	return NewClient(pool, cf.clustersNamespaces.namespaces), nil
+}
+
+func (cf *clientsFactory) GetClustersNamespaces() map[string][]v1.Namespace {
```

Comment:
It seems this function is now being used anywhere. Is this for a specific purpose?

Reply:
nope, removing it.

Reply:
Ops sorry it is. in the Flux Runtime page

Reply:
Ok I think I'm going to blame my IDE for not finding that reference lol